Add setting for adjusting beatmap offset automatically based on last play

frenzibyte opened the pull request:
- Addresses https://github.com/ppy/osu/discussions/30521
- Supersedes https://github.com/ppy/osu/pull/30586

This re-attempts the PR above while also providing UI/UX feedback and supporting adjust offset on quick restarts. @Inconsist's algorithm to handle scores with high unstable rate has been ported across verbatim (see https://github.com/ppy/osu/discussions/30521#discussion-7432380).

frenzibyte committed:
Allow adjusting beatmap offset automatically based on last play

frenzibyte committed:
Add test coverage

peppy commented:
First test, this feels a bit weird:

<img width="511" height="197" alt="osu! 2025-08-12 at 07 11 57" src="https://github.com/user-attachments/assets/7214399f-8b90-441e-9074-1b2a9569b7b6" />

Adjusted by 0 ms suggests a rounding issue either in display or underlying logic.

peppy reviewed (commented):
As commented

frenzibyte committed:
Adjust offset text and fix incorrect display in certain scenarios

frenzibyte committed:
Display calibration button when offset is manually adjusted

frenzibyte committed:
Update test coverage

frenzibyte commented on osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs:
```diff
@@ -142,13 +146,16 @@ protected override void LoadComplete()
 
         private void currentChanged(ValueChangedEvent<double> offset)
         {
-            Scheduler.AddOnce(updateOffset);
```
I forgot to highlight this when opening the PR; This is a mandatory change for automatic offset adjustment to work when the player is quick-retrying.

This is because the entire settings remain hidden during quick-retry, and when this control attempts to automatically adjust offset, the offset change is scheduled here and never gets executed until the next time a "slow" retry is performed and the settings are visible again.

Fixing this differently involves splitting this code to a separate component class that's loaded at a `PlayerLoader` level and remains updated even when settings are hidden, but I attempted that and felt I'm overengineering things.

I don't particularly see any issue with not scheduling this line, so I'm going with it.

peppy replied:
Your explanation above does not align.

Test case `TestAutomaticAdjustment` fails without gameplay or hidden state being involved.

frenzibyte replied:
The test is actually failing due to special logic made to have the calibration button show itself when the user modifies the offset after it was automatically adjusted.

Specifically, the control adjusts the offset first then hides the button, but because the offset change callback is scheduled, the button show call executes after the hide call, and so the test fails.

This is fixed in https://github.com/ppy/osu/pull/34612/commits/5462f0c20ffb4647ccf1769edae0816e5d371ed9 by taking away non-realm-related logic from scheduling. Now scheduling doesn't break any tests.

peppy reviewed (commented):
More issues.

frenzibyte committed:
Add failing test case

frenzibyte commented:
I've also fixed in https://github.com/ppy/osu/pull/34612/commits/5462f0c20ffb4647ccf1769edae0816e5d371ed9 another issue where clicking the button after manually modifing offset from auto adjust does not apply the suggested offset, but applies the pre-UR suggestion instead. I've added test coverage for that in https://github.com/ppy/osu/pull/34612/commits/c349d536e8e3d02b5829d4ff40cd5711a538ce08.

frenzibyte committed:
Move beatmap offset write logic to a separate method

Note that calling `writeOffsetToBeatmap` from `currentChanged` is still intentionally not
scheduled, see https://github.com/ppy/osu/pull/34612#discussion_r2269052138.

This commit is done so that if that call is scheduled, then the logic
above it that update the play graph and visibility state of the button
still don't get scheduled, otherwise the test will fail (as noticed in https://github.com/ppy/osu/pull/34612#discussion_r2269208998).

frenzibyte committed:
Fix calibrating offset with auto-adjust enabled does not account for UR

Covered by test `TestAutomaticAdjustmentWithUnstableRate`.

frenzibyte committed:
Fix current change check not solid enough

peppy commented on osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs:
```diff
@@ -230,10 +237,15 @@ private void scoreChanged(ValueChangedEvent<ScoreInfo?> score)
+
+            bool autoAdjustBeatmapOffset = config.Get<bool>(OsuSetting.AutomaticallyAdjustBeatmapOffset);
+
+            suggestedOffset = computeSuggestedOffset(median, unstableRate, lastPlayBeatmapOffset, proportionalToUnstableRate: autoAdjustBeatmapOffset);
```
Can this method be inlined here again? It's only used once and extracting it out only seems to add complexity.

frenzibyte replied:
Inlining the method here feels like shoving a lot of lines into this one single `scoreChanged` callback method, or at least that was what I had in mind. I'll inline it.

peppy commented on osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs:
```diff
@@ -320,6 +349,21 @@ public void OnReleased(KeyBindingReleaseEvent<GlobalAction> e)
+
+            double offsetAdjustment = median;
+
+            if (proportionalToUnstableRate && unstableRate >= ur_adjustment_cutoff)
```
This is super weird. A global setting changing the algorithm used.

If we're semi-confident about the new UR basis, then let's just apply it everywhere in all cases.

cc @bdach may have thoughts on this.

peppy commented on osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs:
```diff
@@ -142,39 +146,42 @@ protected override void LoadComplete()
+            // Calibration button may be hidden due to automatic offset adjustment, but it should be visible when the user manually adjusts their offset away from the applied suggestion.
+            calibrateFromLastPlayButton?.Show();
+
+            writeOffsetToBeatmap();
```
```suggestion
            Scheduler.AddOnce(writeOffsetToBeatmap);
```

Else you could have two of these run at the same frame which is what we're trying to avoid.

peppy replied:
This **needs to be mentioned inline**.

I would change this without a blink of an eye because it currently goes against standards without any visible reason.

frenzibyte committed:
Add inline comment

smoogipoo commented:
I'll post it again here, but it seems a lot of issues are around this `lastPlayMedian` value. In order to reduce reviewer strain, I suggested the following to keep the value and compute the suggested offset at usage sites instead:

```diff
diff --git a/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs b/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
index 9de35b0c19..1779f3f99a 100644
--- a/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
+++ b/osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs
@@ -40,6 +40,8 @@ public partial class BeatmapOffsetControl : CompositeDrawable, IKeyBindingHandle
 
         private Bindable<ScoreInfo?> lastAppliedScore { get; } = new Bindable<ScoreInfo?>();
 
+        private readonly Bindable<bool> autoAdjustBeatmapOffset = new Bindable<bool>();
+
         public BindableDouble Current { get; } = new BindableDouble
         {
             MinValue = -50,
@@ -61,18 +63,12 @@ public partial class BeatmapOffsetControl : CompositeDrawable, IKeyBindingHandle
         [Resolved]
         private Player? player { get; set; }
 
-        [Resolved]
-        private OsuConfigManager config { get; set; } = null!;
-
+        private double lastPlayMedian;
+        private double lastPlayUnstableRate;
         private double lastPlayBeatmapOffset;
         private HitEventTimingDistributionGraph? lastPlayGraph;
-
-        private double suggestedOffset;
-
         private SettingsButton? calibrateFromLastPlayButton;
-
         private IDisposable? beatmapOffsetSubscription;
-
         private Task? realmWriteTask;
         private ScoreInfo? lastValidScore;
 
@@ -107,9 +103,10 @@ public BeatmapOffsetControl()
         }
 
         [BackgroundDependencyLoader]
-        private void load(SessionStatics statics)
+        private void load(SessionStatics statics, OsuConfigManager config)
         {
             statics.BindWith(Static.LastAppliedOffsetScore, lastAppliedScore);
+            config.BindWith(OsuSetting.AutomaticallyAdjustBeatmapOffset, autoAdjustBeatmapOffset);
         }
 
         protected override void LoadComplete()
@@ -200,7 +197,10 @@ private void scoreChanged(ValueChangedEvent<ScoreInfo?> score)
 
             var hitEvents = score.NewValue.HitEvents;
 
-            if (!(hitEvents.CalculateMedianHitError() is double median))
+            if (hitEvents.CalculateMedianHitError() is not double median)
+                return;
+
+            if (hitEvents.CalculateUnstableRate()?.Result is not double unstableRate)
                 return;
 
             // affecting unstable rate here is used as a substitute of determining if a hit event represents a *timed* hit event,
@@ -237,14 +237,10 @@ private void scoreChanged(ValueChangedEvent<ScoreInfo?> score)
             }
 
             lastValidScore = score.NewValue!;
+            lastPlayMedian = median;
+            lastPlayUnstableRate = unstableRate;
             lastPlayBeatmapOffset = Current.Value;
 
-            double unstableRate = hitEvents.CalculateUnstableRate()!.Result;
-
-            bool autoAdjustBeatmapOffset = config.Get<bool>(OsuSetting.AutomaticallyAdjustBeatmapOffset);
-
-            suggestedOffset = computeSuggestedOffset(median, unstableRate, lastPlayBeatmapOffset, proportionalToUnstableRate: autoAdjustBeatmapOffset);
-
             LinkFlowContainer offsetText;
 
             referenceScoreContainer.AddRange(new Drawable[]
@@ -271,7 +267,7 @@ private void scoreChanged(ValueChangedEvent<ScoreInfo?> score)
                 }
             });
 
-            if (autoAdjustBeatmapOffset && !Current.Disabled)
+            if (autoAdjustBeatmapOffset.Value && !Current.Disabled)
             {
                 bool offsetChanged = applySuggestedOffset();
 
@@ -293,7 +289,7 @@ private bool applySuggestedOffset()
         {
             double lastOffset = Current.Value;
 
-            Current.Value = suggestedOffset;
+            Current.Value = computeSuggestedOffset(lastPlayMedian, lastPlayUnstableRate, lastPlayBeatmapOffset, proportionalToUnstableRate: autoAdjustBeatmapOffset.Value);
             lastAppliedScore.Value = lastValidScore;
 
             return !Precision.AlmostEquals(Current.Value, lastOffset, Current.Precision / 2);
@@ -315,7 +311,10 @@ protected override void Update()
             bool allow = allowOffsetAdjust;
 
             if (calibrateFromLastPlayButton != null)
+            {
+                double suggestedOffset = computeSuggestedOffset(lastPlayMedian, lastPlayUnstableRate, lastPlayBeatmapOffset, proportionalToUnstableRate: autoAdjustBeatmapOffset.Value);
                 calibrateFromLastPlayButton.Enabled.Value = allow && !Precision.AlmostEquals(suggestedOffset, Current.Value, Current.Precision / 2);
+            }
 
             Current.Disabled = !allow;
         }

```

frenzibyte committed:
Adjust code to be more reviewable

frenzibyte commented:
Applied, thanks.

peppy committed:
Always use UR algorithm

peppy committed:
Reorder class to suck a bit less

Inconsist commented:
Just a reminder that the last time when the ur algorithm was used universally, the prompt would still say something like:

“average hit error.         10.5ms early”

And when you hit the button it would adjust the offset to some seemingly random value like +5.8ms which would make no sense to the player. So the offset in the prompt would have to be the ur based value too.

peppy commented:
It's done correctly to match this time. Thanks for your feedback.
